### PR TITLE
Add levelbuilder map editor for Star Wars BB-8 levels

### PR DESCRIPTION
**Pull request by joshlory**
screen shot 2017-05-15 at 6 32 18 pm

TODO:
- [x] Fix starting map tiles.
- [x] Existing issue: after changing the skin, need to save before the new map editor is rendered.

### CODE REVIEW

**Comment by joshlory** [comment]
Should I create a `StarWarsGrid` and `StarWarsEvents` subtype of `Studio`?  Subtype of `Grid`?  Trying to figure out the best way to organize these concepts in Level Builder.

+cc @breville @kiki-lee @ryansloan

File: dashboard/app/views/levels/editors/_grid.html.haml [hunk 1]
```diff
@@ -5,9 +5,9 @@
   - if @level.is_a? Karel
     #gridEditor
     :javascript
-      var maze = eval(#{@level.properties['maze'].try(:to_json)});
-      var initialDirt = eval(#{@level.properties['initial_dirt'].try(:to_json)});
-      var serializedMaze = eval(#{@level.properties['serialized_maze'].try(:to_json)});
+      var maze = JSON.parse(#{@level.properties['maze'].try(:to_json) || 'null'});
+      var initialDirt = JSON.parse(#{@level.properties['initial_dirt'].try(:to_json) || 'null'});
+      var serializedMaze = JSON.parse(#{@level.properties['serialized_maze'].try(:to_json) || 'null'});
```
Comment (Hamms) [inline dashboard/app/views/levels/editors/_grid.html.haml:10]: :+1:

File: apps/src/code-studio/components/Grid.jsx [hunk 2]
```diff
@@ -1,7 +1,10 @@
 import React from 'react';
-import { SquareType } from '@cdo/apps/studio/constants';
+import { SquareType, WallTypeMask, WallCoordRowMask, WallCoordRowShift, WallCoordColMask, WallCoordColShift } from '@cdo/apps/studio/constants';
```
Comment (Hamms) [inline apps/src/code-studio/components/Grid.jsx]: nit: [our styleguide would prefer these be on separate lines](https://github.com/airbnb/javascript#modules--multiline-imports-over-newlines)

File: apps/src/code-studio/components/StarWarsGridCellEditor.jsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  covered_crates_3b: 0x350000,
+};
+
+export default class StarWarsGridCellEditor extends React.Component {
+  static propTypes = {
+    cell: React.PropTypes.object.isRequired,
+    row: React.PropTypes.number.isRequired,
+    col: React.PropTypes.number.isRequired,
+    onUpdate: React.PropTypes.func.isRequired,
+  }
```
Comment (Hamms) [inline apps/src/code-studio/components/StarWarsGridCellEditor.jsx:27]: do we support this ES7 method of propTypes declaration?

File: apps/src/code-studio/components/StarWarsGridCellEditor.jsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
+import React from 'react';
+import { WallTypeMask, WallTypeShift } from '@cdo/apps/studio/constants';
+import tileGuide from '../../../static/code_studio/tile-guide.png';
```
Comment (Hamms) [inline apps/src/code-studio/components/StarWarsGridCellEditor.jsx:3]: 😮 I didn't know we could do this

File: apps/src/code-studio/components/StarWarsGridCellEditor.jsx (hunk 3 above)
Comment (joshlory) [reply apps/src/code-studio/components/StarWarsGridCellEditor.jsx:27]: Yep!  An example: https://github.com/code-dot-org/code-dot-org/blob/011339e5cf04e5c59b886c6731add5ab6483c9b3/apps/src/code-studio/components/stageExtras/StageExtras.jsx#L6-L9

Comment (Hamms) [reply apps/src/code-studio/components/StarWarsGridCellEditor.jsx:27]: 🎉

**Comment by breville** [comment]
fwiw, I don't have any particular feelings on how to organise the level types, mainly because I don't know the broader implications.

File: dashboard/app/models/levels/star_wars_grid.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+class StarWarsGrid < Studio
+  def self.create_from_level_builder(params, level_params)
+    level = new(
+      level_params.merge(
+        user: params[:user],
+        game: Game.custom_studio,
+        level_num: 'custom',
+        start_blocks: '<xml><block type="when_run"/></xml>',
+        wall_map_collisions: true,
+        block_moving_into_walls: true,
```
Comment (joshlory) [inline dashboard/app/models/levels/star_wars_grid.rb:36]: These options will go away with @balderdash's change to enable these whenever a collision map is present.

**Pull request merged**